src/hyperlane-adapter/HyperlaneReceiverAdapter.sol

```diff
@@ -0,0 +1,185 @@
+   */
+  error MessageIdAlreadyExecuted(bytes32 messageId);
+
+  /**
+   * @notice Emitted when a sender adapter for a source chain is updated.
+   * @param srcChainId Source chain identifier.
+   * @param senderAdapter Address of the sender adapter.
+   */
+  event SenderAdapterUpdated(uint256 srcChainId, IMessageDispatcher senderAdapter);
+
+  /* Constructor */
+  /**
+   * @notice HyperlaneReceiverAdapter constructor.
+   * @param _mailbox Address of the Hyperlane `Mailbox` contract.
+   */
+  constructor(address _mailbox) {
+    if (_mailbox == address(0)) {
+      revert Errors.InvalidMailboxZeroAddress();
+    }
+    mailbox = IMailbox(_mailbox);
+  }
+
+  /// @notice Restrict access to trusted `Mailbox` contract.
+  modifier onlyMailbox() {
+    if (msg.sender != address(mailbox)) {
+      revert Errors.UnauthorizedMailbox(msg.sender);
+    }
+    _;
+  }
+
+  /// @inheritdoc ISpecifiesInterchainSecurityModule
+  function interchainSecurityModule() external view returns (IInterchainSecurityModule) {
+    return ism;
+  }
+
+  /**
+   * @notice Sets the ISM for this adapter/recipient.
+   * @param _ism The ISM contract address.
+   */
+  function setIsm(address _ism) external onlyOwner {
+    ism = IInterchainSecurityModule(_ism);
+    emit IsmSet(_ism);
+  }
+
+  function executeMessage(
+    address _to,
+    bytes memory _message,
+    bytes32 _messageId,
+    uint256 _fromChainId,
+    address _from,
+    bool _executedMessageId
+  ) internal {
+    MessageLib.executeMessage(_to, _message, _messageId, _fromChainId, _from, _executedMessageId);
+
+    emit MessageIdExecuted(_fromChainId, _messageId);
+  }
+
+  function executeMessageBatch(
+    MessageLib.Message[] memory _messages,
+    bytes32 _messageId,
+    uint256 _fromChainId,
+    address _from,
+    bool _executedMessageId
+  ) internal {
+    MessageLib.executeMessageBatch(_messages, _messageId, _fromChainId, _from, _executedMessageId);
+
+    emit MessageIdExecuted(_fromChainId, _messageId);
+  }
+
+  /**
+   * @notice Called by Hyperlane `Mailbox` contract on destination chain to receive cross-chain messages.
+   * @dev _origin Source chain domain identifier (not currently used).
+   * @param _sender Address of the sender on the source chain.
+   * @param _body Body of the message.
+   */
+  function handle(
+    uint32,
+    /* _origin*/
+    bytes32 _sender,
+    bytes memory _body
+  ) external onlyMailbox {
+    address adapter = TypeCasts.bytes32ToAddress(_sender);
+    bool _executedMessageId;
+    (
+      MessageLib.Message[] memory _messages,
+      bytes32 msgId,
+      uint256 srcChainId,
+      address srcSender
+    ) = abi.decode(_body, (MessageLib.Message[], bytes32, uint256, address));
+
+    if (IMessageDispatcher(adapter) != senderAdapters[srcChainId]) {
+      revert Errors.UnauthorizedAdapter(srcChainId, adapter);
+    }
+    if (executedMessages[msgId]) {
+      revert MessageIdAlreadyExecuted(msgId);
+    } else {
```

Comment:
This else is technically superflous, i like having all the revert conditions up front, and then do the state changes after

Reply:
[Fixed](https://github.com/pooltogether/ERC5164/pull/17/commits/9bdd08bf464430ce67042308eac41cfd2865fee5#r1385022171)